keps/sig-storage/20191213-csi-inline-volume-size.md

```diff
@@ -0,0 +1,191 @@
+
+- [ ] kubernetes/enhancements issue in release milestone, which links
+      to KEP (this should be a link to the KEP location in
+      kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in
+      [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents,
+      links to mailing list discussions/SIG meetings, relevant
+      PRs/issues, release notes
+
+## Summary
+
+The [CSIVolumeSource API in Kubernetes
+1.16](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#csivolumesource-v1-core)
+for [CSI ephemeral inline
+volumes](https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html)
+currently supports only two standardized parameters for the volume:
+- `readOnly`
+- `fsType`
+
+All other parameters for the volume must be passed as
+`volumeAttributes` with key/value pairs that are driver-specific and
+thus opaque to Kubernetes. This API was chosen because it directly
+maps to parameters supported by the underlying
+[NodePublishVolumeRequest](https://github.com/container-storage-interface/spec/blob/4731db0e0bc53238b93850f43ab05d9355df0fd9/lib/go/csi/csi.pb.go#L3604-L3652).
+
+One common questions that developers and users have is how the size of
+the volume can be specified. The proposal is to introduce a
+standardized `fsSize` parameter as answer to that and gradually
+replace driver-specific parameters that may have been used before.
+
+## Motivation
+
+### Goals
+
+- Add a `CSIVolumeSource.fsSize` field.
+- Pass it to drivers as additional entry in `NodePublishVolumeRequest.VolumeContext`
+  if the driver has opted into getting additional fields there.
+
+### Non-Goals
+
+- Validate whether a CSI driver really supports this field. This is
+  consistent with `readOnly` and `fsType` which also may be silently
+  ignored.
+
+## Proposal
+
+### User Stories
+
+#### Ephemeral local volumes
+
+Local volumes are useful as scratch
+space. [PMEM-CSI](https://github.com/intel/pmem-csi) and
+[TopoLVM](https://github.com/cybozu-go/topolvm/blob/master/README.md)
+are two examples for CSI drivers which dynamically create volumes of a
+fixed capacity and thus need a size parameter.
+
+### Capacity-aware scheduling of pods
+
+Scheduling pods with ephemeral inline volumes onto nodes [with
+sufficient free storage
+capacity](https://github.com/kubernetes/enhancements/pull/1353)
+depends (among other information, like free capacity) on knowing the
+size of the volumes. If the size is only specified via some
+vendor-specific parameter, it's not available to the Kubernetes
+scheduler.
+
+### fsSize field
```

Comment:
Dose the scheduler have to look at this field and compare it with the storage pool capacity reported in #1353?

Reply:
Yes, that was the original motivation for adding this field.